### PR TITLE
feat: add harvester_cloudinit_secret

### PR DESCRIPTION
**Pull request by FrankYang0529**
Issue: https://github.com/harvester/harvester/issues/2958

## Test plan

**Setup**
Follow [PR Test Env Setup](https://github.com/harvester/terraform-provider-harvester/wiki/PR-Test-Env-Setup)

**Case 1: Apply same terraform doesn't trigger change**
1. Create a file `main.tf`.
```
resource "harvester_cloudinit_secret" "ex1" {
  name             = "ex1"
  namespace        = "default"
  user_data_base64 = "I2Nsb3VkLWNvbmZpZwpwYWNrYWdlX3VwZGF0ZTogdHJ1ZQpwYWNrYWdlczoKICAtIHFlbXUtZ3Vlc3QtYWdlbnQKcnVuY21kOgogIC0gLSBzeXN0ZW1jdGwKICAgIC0gZW5hYmxlCiAgICAtIC0tbm93CiAgICAtIHFlbXUtZ3Vlc3QtYWdlbnQuc2VydmljZQpwYXNzd29yZDogdGVzdApjaHBhc3N3ZDoKICBleHBpcmU6IGZhbHNlCnNzaF9wd2F1dGg6IHRydWUK"
}

resource "harvester_cloudinit_secret" "ex2" {
    name = "ex2"
    namespace = "default"

    user_data    = <<-EOF
    #cloud-config
    password: test
    chpasswd:
      expire: false
    ssh_pwauth: true
    package_update: true
    packages:
      - qemu-guest-agent
    runcmd:
      - - systemctl
        - enable
        - '--now'
        - qemu-guest-agent
    EOF
    network_data = ""
}
```
2. Run `terraform apply`. Terraform should add `ex1` and `ex2` secrets to the cluster.
3. Run `terraform apply` again. Nothing change.

**Case 2: Change content to empty will update the secret**
1. Follow case 1.
2. Update the file `main.tf`.
```
resource "harvester_cloudinit_secret" "ex1" {
  name             = "ex1"
  namespace        = "default"
  user_data_base64 = ""
}

resource "harvester_cloudinit_secret" "ex2" {
    name = "ex2"
    namespace = "default"

    user_data    = ""
    network_data = ""
}
```
3. Run `terraform apply`. The content of `ex1` and `ex2` secrets will be empty.
4. Run `terraform apply` again. Nothing change.

### CODE REVIEW

**Review by futuretea — COMMENTED** [review]
Also need to consider another test case：
1. apply with user_data not empty
2. change user_data to empty
3. apply again

**Comment by FrankYang0529** [comment]
> Also need to consider another test case：
> 
> 1. apply with user_data not empty
> 2. change user_data to empty
> 3. apply again

Added test case 2. Thanks.

**Comment by iosifnicolae2** [comment]
Any idea when this PR will be merged?

File: internal/provider/cloudinitsecret/resource_cloudinitsecret.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+		stateGetter.States[constants.FieldCloudInitSecretUserData] = string(userdata)
+		stateGetter.States[constants.FieldCloudInitSecretUserDataBase64] = ""
+	}
+	if d.Get(constants.FieldCloudInitSecretNetworkData) != "" && d.Get(constants.FieldCloudInitSecretNetworkDataBase64) == "" {
+		networkdata, err := base64.StdEncoding.DecodeString(stateGetter.States[constants.FieldCloudInitSecretNetworkDataBase64].(string))
+		if err != nil {
+			return fmt.Errorf("can't decode userdata: %v", err)
+		}
+		stateGetter.States[constants.FieldCloudInitSecretUserData] = string(networkdata)
+		stateGetter.States[constants.FieldCloudInitSecretUserDataBase64] = ""
```
Comment (chrisho) [inline internal/provider/cloudinitsecret/resource_cloudinitsecret.go:127]: Should this be network data?

File: internal/provider/cloudinitsecret/resource_cloudinitsecret.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		if err != nil {
+			return fmt.Errorf("can't decode userdata: %v", err)
+		}
+		stateGetter.States[constants.FieldCloudInitSecretUserData] = string(userdata)
+		stateGetter.States[constants.FieldCloudInitSecretUserDataBase64] = ""
+	}
+	if d.Get(constants.FieldCloudInitSecretNetworkData) != "" && d.Get(constants.FieldCloudInitSecretNetworkDataBase64) == "" {
+		networkdata, err := base64.StdEncoding.DecodeString(stateGetter.States[constants.FieldCloudInitSecretNetworkDataBase64].(string))
+		if err != nil {
+			return fmt.Errorf("can't decode userdata: %v", err)
```
Comment (chrisho) [inline internal/provider/cloudinitsecret/resource_cloudinitsecret.go:124]: ditto

**Review by chrisho — CHANGES_REQUESTED** [review]

File: internal/provider/cloudinitsecret/resource_cloudinitsecret.go (hunk 2 above)
Comment (chrisho) [inline internal/provider/cloudinitsecret/resource_cloudinitsecret.go:124]: Should this be network data?

File: internal/provider/cloudinitsecret/resource_cloudinitsecret.go (hunk 1 above)
Comment (chrisho) [inline internal/provider/cloudinitsecret/resource_cloudinitsecret.go:127]: ditto

File: internal/provider/cloudinitsecret/resource_cloudinitsecret.go (hunk 2 above)
Comment (futuretea) [reply internal/provider/cloudinitsecret/resource_cloudinitsecret.go]: Fixed. Thanks!

**Review by chrisho — APPROVED** [review]
LGTM, the test cases are working fine, thanks.

**Review by futuretea — APPROVED** [review]

**Review by guangbochen — APPROVED** [review]
LGTM, and verified that the testing case works.

**Comment by guangbochen** [comment]
@bk201 please take a look when u are back only; merge the PR to process the validation first, thanks.

**Pull request merged**